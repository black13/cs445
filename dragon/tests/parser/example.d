program example(input, output);
var x, y: integer;
function gcd(a, b: integer): integer;
begin
    if b = 0 then gcd := a
    else gcd := gcd(b, a mod b)
end;

begin
    scanline(x, y);
    putall(gcd(x, y))
end.
