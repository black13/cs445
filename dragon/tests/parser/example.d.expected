PROGRAM `example` WITH ARGS:
  `input`
  `output`
AND VARIABLES:
  DECLARATION WITH TYPE:
    INTEGER
  OF NAMES:
    `x`
    `y`
AND SUBPROGRAMS:
  FUNCTION `gcd` WITH ARGS:
    DECLARATION WITH TYPE:
      INTEGER
    OF NAMES:
      `a`
      `b`
  RETURNING:
    INTEGER
  WITH SUBPROGRAMS:
  AND VARIABLES
  DOES:
    IF FOLLOWING IS TRUE:
      BINOP EQ
        LEFT:
          b
        RIGHT:
          LIT `0`
    THEN:
      ASSIGN
        gcd
      TO
        a
    OTHERWISE:
      ASSIGN
        gcd
      TO
        APPLY `gcd`
        TO ARGS:
          b
          BINOP MOD
            LEFT:
              a
            RIGHT:
              b
DOES:
  APPLY `read` WITH:
    x
    y
  APPLY `write` WITH:
    APPLY `gcd`
    TO ARGS:
      x
      y
-- done dumping ast --
